### PR TITLE
241 files to process class

### PR DESCRIPTION
**Pull request by emlynjdavies**
This PR adds a new feature to the `pyopia process`command that can chuck the list of raw files in a way that also provides the correct background files to each chunk (depending on the config settings for no background, fixed background, and moving background) and process each chunk in parallel using threading.

The new optional input --num-chunks can be used to specify how many chunks (and ergo threads) to use:

```
Usage: pyopia process [OPTIONS] CONFIG_FILENAME                                                                                                                        
                                                                                                                                                                        
 Run a PyOPIA processing pipeline based on given a config.toml                                                                                                          
 Parameters ---------- config_filename : str     config filename numchunks : int, optional     split the dataset into chucks, and process in parallell, by default 1    
                                                                                                                                                                        
╭─ Arguments ──────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────╮
│ *    config_filename      TEXT  [default: None] [required]                                                                                                           │
╰──────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────╯
╭─ Options ────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────╮
│ --num-chunks        INTEGER  [default: 1]                                                                                                                            │
│ --help                       Show this message and exit.                                                                                                             │
╰──────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────╯
```

Example:

```
pyopia process congif.toml --num-chunks 8
```

You must use append = false mode in the output step if you want to use chunking.

The previous functionality remains unaffected and `pyopia process config.toml`works as is did before (without threading or chunking)

### CODE REVIEW

File: pyopia/pipeline.py [hunk 1]
```diff
@@ -322,3 +324,79 @@ def build_steps(toml_steps):
...
+        '''
+        assert chunks > 0, 'you must have at least one chunk'
+        n = int(np.ceil(len(self.files) / chunks))
+        self.chunked_files = [self.files[i:i + n] for i in range(0, len(self.files), n)]
+
+    def insert_bg_files_into_chunks(self, bgshift_function='pass'):
+        average_window = len(self.background_files)
+        for i, c in enumerate(self.chunked_files):
+            if i > 0 and bgshift_function != 'pass':
+                self.background_files.extend(self.chunked_files[i-1][-average_window:])
```
Comment (nordam) [inline pyopia/pipeline.py:388]: Not obvious to me what is happening here. What happens if ```bgshift_function == 'pass'```? And why skip the first image in the loop (```i > 0```)?

Perhaps add a comment?

File: pyopia/cli.py [hunk 2]
```diff
@@ -105,23 +106,34 @@ def generate_config(instrument: str, raw_files: str, model_path: str, outfolder:
 
 
 @app.command()
-def process(config_filename: str):
+def process(config_filename: str, chunks: int = 1):
```
[the remark this answers is not in the document]
Comment (emlynjdavies) [reply pyopia/cli.py:109]: what about numchunks?

the number of threads is inferred from the number of chunks

**Review by nordam — COMMENTED** [review]
Left a couple of minor comments, think it overall makes sense

File: pyopia/cli.py [hunk 3]
```diff
@@ -179,15 +204,25 @@ def setup_logging(pipeline_config):
...
-    log_format = '%(asctime)s %(levelname)s [%(module)s.%(funcName)s] %(message)s'
-    logging.basicConfig(level=log_level, format=log_format, filename=log_file,
-                        datefmt='%Y-%m-%d %H:%M:%S')
+    log_format = '%(asctime)s %(levelname)s %(threadName)s [%(module)s.%(funcName)s] %(message)s'
+    logging.basicConfig(level=log_level, datefmt='%Y-%m-%d %H:%M:%S', format=log_format, handlers=handlers)
+
+
+def check_chunks(chunks, pipeline_config):
+    if chunks < 1:
+        raise RuntimeError('You must have at least 1 chunk')
```
[the remark this answers is not in the document]
Comment (emlynjdavies) [reply pyopia/cli.py:221]: this is enforced by the input to process()

File: pyopia/pipeline.py [hunk 4]
```diff
@@ -322,3 +324,88 @@ def build_steps(toml_steps):
...
+        Parameters
+        ----------
+        path_to_filelist : str
+            Path to txt file to write
+        '''
+        with open(path_to_filelist, 'w') as fh:
+            [fh.writelines(L + '\n') for L in self.files]
+
+    def prepare_chunking(self, chunks, average_window, bgshift_function):
+        if chunks > len(self.files) // 2:
```
[the remark this answers is not in the document]
Comment (emlynjdavies) [reply pyopia/pipeline.py:366]: if num_chunks == 1 then it will behave the same as not chunking (i.e. one list of files)

**Review by nordam — APPROVED** [review]

**Pull request merged**